parsec/core/gui/keys_widget.py

```diff
@@ -0,0 +1,98 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+import shutil
+import os
+from pathlib import Path
+
+from PyQt5.QtCore import pyqtSignal
+from PyQt5.QtWidgets import QWidget, QFileDialog
+
+from parsec.core.local_device import (
+    list_available_devices,
+    AvailableDevice,
+    load_device_file,
+    get_devices_dir,
+)
+
+from parsec.core.gui.ui.keys_widget import Ui_KeysWidget
+from parsec.core.gui.ui.key_widget import Ui_KeyWidget
+from parsec.core.gui.custom_dialogs import show_error, ask_question
+from parsec.core.gui.lang import translate
+
+
+class KeyWidget(QWidget, Ui_KeyWidget):
+    export_clicked = pyqtSignal(AvailableDevice)
+
+    def __init__(self, device, parent=None):
+        super().__init__(parent=parent)
+        self.device = device
+        self.setupUi(self)
+        self.label_org.setText(device.organization_id)
+        self.label_device.setText(device.device_label)
+        self.label_user.setText(device.human_handle.label)
+        self.export_button.clicked.connect(self._on_export)
+
+    def _on_export(self):
+        self.export_clicked.emit(self.device)
+
+
+class KeysWidget(QWidget, Ui_KeysWidget):
+    def __init__(self, config, parent):
+        super().__init__(parent=parent)
+        self.config = config
+        self.setupUi(self)
+        self.reload_devices()
+        self.button_import_key.clicked.connect(self._on_import_key)
+
+    def reload_devices(self):
+        layout = self.scroll_content.layout()
+        for _ in range(layout.count()):
+            item = self.scroll_content.layout().takeAt(0)
+            layout.removeItem(item)
+            w = item.widget()
+            if w is not None:
+                w.setParent(None)
+        devices = list_available_devices(self.config.config_dir)
+        for device in devices:
+            w = KeyWidget(device, parent=self)
+            w.export_clicked.connect(self._on_export_key)
+            self.scroll_content.layout().insertWidget(self.scroll_content.layout().count() - 1, w)
+
+    def _on_export_key(self, device):
+        file_dialog = QFileDialog()
+        output_directory = file_dialog.getExistingDirectory()
+        if not output_directory:
+            return
+        _, key_name = os.path.split(device.key_file_path)
+        try:
+            shutil.copyfile(device.key_file_path, os.path.join(output_directory, key_name))
```

Comment:
Good idea, i did not find a way to set a default name and let the user specify the name. I am still using `QFileDialog.getExistingDirectory()`

Reply:
You can let the user specify the name using `QFileDialog.getSaveFileName()` (https://doc.qt.io/qt-5/qfiledialog.html#getSaveFileName). The `dir` argument is not well named but it can take a whole path by default. For example:
`save_path, _ = QFileDialog.getSaveFileName(self, "Title", "/home/max/parsec-myorg-max-debian.keys")`
will open a dialog where the user will be able to chose where to save the keys (defaulting to his home directory), with a default name of "parsec-myorg-max-debian.keys" that the user will be able to change.